pkg/kv/kvserver/gc/gc.go

```diff
@@ -390,140 +457,481 @@ func processReplicatedKeyRange(
+			if b.clearRangeEndKey.Compare(lastKey) > 0 {
+				// If flushed batch was split by start of the range, then we must remove
+				// preceding part from clear range total to avoid.
+				b.clearRangeCounters.sub(update)
+				b.clearRangeCounters.add(b.partialPointBatchCounters)
+				b.clearRangeEndKey = lastKey[:len(lastKey):len(lastKey)].Next()
 			}
-			batchGCKeys = nil
-			batchGCKeysBytes = 0
-			alloc = bufalloc.ByteAllocator{}
+			b.partialPointBatchCounters = gcBatchCounters{}
 		}
 	}
-	// We need to send out last intent cleanup batch.
-	if err := intentBatcher.maybeFlushPendingIntents(ctx); err != nil {
-		if errors.Is(err, ctx.Err()) {
-			return false, err
+
+	if b.clearRangeEnabled {
+		if b.prevWasNewest {
+			if key == nil {
+				// We reuse the start key slice to avoid reallocating it on every new
+				// non-garbage key.
+				key = append(b.clearRangeStartKey.Key[:0], cur.key.Key...)
+			}
+			b.clearRangeStartKey = storage.MVCCKey{Key: key}
+			b.clearRangeCounters.keysAffected++
 		}
-		log.Warningf(ctx, "failed to cleanup intents batch: %v", err)
+		b.clearRangeStartKey.Timestamp = cur.key.Timestamp
+		b.clearRangeCounters.keyBytes += keySize
+		b.clearRangeCounters.valBytes += int64(cur.mvccValueLen)
+		b.clearRangeCounters.versionsAffected++
 	}
-	if len(batchGCKeys) > 0 {
-		if err := gcer.GC(ctx, batchGCKeys, nil, nil); err != nil {
-			return false, err
+
+	b.prevWasNewest = isNewestPoint
+	return counterUpdate, nil
+}
+
+// maybeFlushPendingBatches flushes accumulated GC requests. If clear range is
+// used and current data is eligible for deletion then pending points batch and
+// clear range request are flushed.
+// If there's not enough point keys to justify clear range, then point point
+// batches may be flushed:
+//   - if there's a single batch and it didn't reach desired size, nothing is
+//     flushed
+//   - if there's more than single points batch accumulated, then all batches
+//     except for last are flushed
+func (b *gcKeyBatcher) maybeFlushPendingBatches(
+	ctx context.Context,
+) (counterUpdate gcInfoUpdate, err error) {
+	// Find where in first points batch is start key and flush "prefix".
+	if b.clearRangeEnabled && b.clearRangeCounters.versionsAffected >= b.clearRangeMinKeys {
+		// Optionally flush parts of the first batch if it doesn't match
+		// end of the range.
+		batchLen := len(b.pointsBatches[0].batchGCKeys)
+		// Find a key that is equal or less than end of clear range key.
+		// Batch is sorted in the reverse order, timestamps are not relevant since
+		// we can't have the same key more than once within single points batch.
+		lastIdx := sort.Search(batchLen, func(i int) bool {
+			return b.pointsBatches[0].batchGCKeys[i].Key.Compare(b.clearRangeEndKey) < 0
+		})
+		// If lastIdx == 0 we can have a clear range request which goes back to
+		// previous key completely covering this range. If this is true then we don't
+		// need to handle this batch.
```

Comment:
Should we tighten the `EndKey` in this case to avoid dropping an unnecessarily wide clearrange? Could that happen automatically whenever we're flushing a point key batch?